src/theory/bv/bvgauss.h

```diff
@@ -0,0 +1,150 @@
+   *                 - bit-vector constants
+   *                 - extracts or zero extensions on bit-vector constants
+   *                 - of arbitrary nesting level
+   *               and p is co-prime to all bit-vector constants for which
+   *               a multiplicative inverse has to be computed.
+   *
+   *   UNIQUE  ... determined for all unknowns, e.g., x = 4
+   *
+   *   PARTIAL ... e.g., x = 4 - 2z
+   *
+   *   NONE    ... no solution
+   *
+   *   Given a matrix A representing an equation system, the resulting
+   *   matrix B after applying GE represents, e.g.:
+   *
+   *   B = 1 0 0 2 <-    UNIQUE
+   *       0 1 0 3 <-
+   *       0 0 1 1 <-
+   *
+   *   B = 1 0 2 4 <-    PARTIAL
+   *       0 1 3 2 <-
+   *       0 0 1 1
+   *
+   *   B = 1 0 0 1       NONE
+   *       0 1 1 2
+   *       0 0 0 2 <-
+   */
+  enum class Result
+  {
+    INVALID,
+    UNIQUE,
+    PARTIAL,
+    NONE
+  };
+
+  /**
+   * Determines if an overflow may occur in given 'expr'.
+   *
+   * Returns 0 if an overflow may occur, and the minimum required
+   * bit-width such that no overflow occurs, otherwise.
+   *
+   * Note that it would suffice for this function to be Boolean.
+   * However, it is handy to determine the minimum required bit-width for
+   * debugging purposes.
+   */
+  static unsigned getMinBwExpr(Node expr);
+
+  /**
+   * Apply Gaussian Elimination on a set of equations modulo some (prime)
+   * number given as bit-vector equations.
+   *
+   * IMPORTANT: Applying GE modulo some number (rather than modulo 2^bw)
+   * on a set of bit-vector equations is only sound if this set of equations
+   * has a solution that does not produce overflows. Consequently, we only
+   * apply GE if the given bit-width guarantees that no overflows can occur
+   * in the given set of equations.
+   *
+   * Note that the given set of equations does not have to be modulo a prime
+   * but can be modulo any arbitrary number. However, if it is indeed modulo
+   * prime, GE is guaranteed to succeed, which is not the case, otherwise.
+   *
+   * Returns INVALID if GE can not be applied, UNIQUE and PARTIAL if GE was
+   * successful, and NONE, otherwise.
+   */
+  static Result gaussElimRewriteForUrem(
+      std::vector<Node>& equations,
+      std::unordered_map<Node, Node, NodeHashFunction>& res);
+
+  /**
+   * Apply Gaussian Elimination modulo a (prime) number.
+   * The given equation system is represented as a matrix of Integers.
+   *
+   * Note that given 'prime' does not have to be prime but can be any
+   * arbitrary number. However, if 'prime' is indeed prime, GE is guaranteed
+   * to succeed, which is not the case, otherwise.
+   *
+   * Returns INVALID if GE can not be applied, UNIQUE and PARTIAL if GE was
+   * successful, and NONE, otherwise.
+   */
+  static Result gaussElim(Integer prime,
+                          const std::vector<Integer>& rhs,
+                          const std::vector<std::vector<Integer>>& lhs,
```

Comment:
This was for debugging purposes and I'd prefer to keep it that way since it doesn't cost much (if I have to revisit, which I just had to, I might have to reintroduce this for debugging).

Reply:
Would it be an option to do the copy outside of gaussElim if you need it for debugging purposes? I agree that it probably is not a huge performance issue, so I am fine leaving it in if need be.

Reply:
Got rid of result matrix, result is now stored in given matrix.